llmfoundry/data/utils.py

```diff
@@ -20,6 +20,55 @@
 log = logging.getLogger(__name__)
 
 
+class LossGeneratingTokensCollatorWrapper:
+    """Collator wrapper to add loss generating token counts to batch."""
+
+    def __init__(
+        self,
+        base_collator: Callable,
+        token_counting_func: Callable[[Batch], Union[int, dict[str, int]]],
+    ):
+        self.base_collator = base_collator
+        self.token_counting_func = token_counting_func
+
+    def __call__(self, examples: list[Any]) -> dict[str, torch.Tensor]:
+        batch = self.base_collator(examples)
+
+        # Add token counts to batch as a list, one for each row, so that microbatch splitting works
+        output = {
+            'total_tokens': [],
+            'loss_generating_tokens': [],
+        }
+        num_rows = batch['input_ids'].shape[0]
+        for row in range(num_rows):
+            row_batch = {
+                'input_ids': batch['input_ids'][row].unsqueeze(0),
+            }
+            if 'attention_mask' in batch:
+                row_batch['attention_mask'] = batch['attention_mask'][
+                    row].unsqueeze(0)
+            if 'labels' in batch:
+                row_batch['labels'] = batch['labels'][row].unsqueeze(0)
+            if 'decoder_attention_mask' in batch:
+                row_batch['decoder_attention_mask'] = batch[
+                    'decoder_attention_mask'][row].unsqueeze(0)
```

Comment:
can we instead modify token_counting_func? It feels awkward to have a side-effect on other attributes here when it should just be appending new data